pysmt/substituter.py

```diff
@@ -15,55 +15,54 @@
 #   limitations under the License.
 #
+import warnings
+
 from six import iteritems
 
 import pysmt.walkers
 import pysmt.operators as op
 
+
 class Substituter(pysmt.walkers.DagWalker):
+    """Performs substitution of a set of terms within a formula.
 
-    def __init__(self, env):
-        pysmt.walkers.DagWalker.__init__(self, env=env, invalidate_memoization=True)
-        self.manager = self.env.formula_manager
+    Let f be a formula ans subs be a map from formula to formula.
+    Substitution returns a formula f' in which all occurrences of the
+    keys of the map have been replaced by their value.
 
-        self.set_function(self.walk_identity, op.SYMBOL, op.REAL_CONSTANT,
-                          op.INT_CONSTANT, op.BOOL_CONSTANT, op.BV_CONSTANT)
+    There are a few considerations to take into account:
+      - In which order to apply the substitution
+      - How to deal with quantified subformulas
 
+    The order in which we apply the substitutions gives raise to two
+    different approaches: Most General Substitution and Most Specific
+    Substitution. Lets consider the example:
 
-    def substitute(self, formula, subs):
-        """Given a map ``subs`` of substitutions, replaces any subformula in
-        formula with the corresponding definition in subs.
-        """
-        # Check that formula is a term
-        if not formula.is_term():
-            raise TypeError("substitute() can only be used on terms.")
+      f = (a & b)
+      subs = {a -> c, (c & b) -> d, (a & b) -> c}
 
-        for (i, k) in enumerate(subs):
-            v = subs[k]
-            # Check that substitutions are terms
-            if not k.is_term():
-                raise TypeError(
-                    "Only terms should be provided as substitutions." +
-                    " Non-term '%s' found." % k)
-            if not v.is_term():
-                raise TypeError(
-                    "Only terms should be provided as substitutions." +
-                    " Non-term '%s' found." % v)
-            # Check that substitutions belong to the current formula manager
-            if k not in self.manager:
-                raise TypeError(
-                    "Key %d does not belong to the Formula Manager." % i)
-            if v not in self.manager:
-                raise TypeError(
-                    "Value %d does not belong to the Formula Manager." % i)
+    With the Most General Substitution (MGS) we obtain:
+      f' = c
+    with the Most Specific Substitution (MSS) we obtain:
+      f' = d
 
-        return self.walk(formula, substitutions=subs)
+    The default behavior before version 0.5 was MSS. However, this
+    leads to unexpected results when dealing with literals, i.e.,
+    substitutions in which both x and Not(x) appear, do not work as
+    expected.  In case of doubt, it is recommended to issue two
+    separate calls to the substitution procedure.
 
+    Quantifiers: ???
```

Comment:
Yes, that was meant to remind me of the issue we discussed... I should have removed it :smile: